tracer/src/Datadog.Trace/Ci/Agent/CIAgentlessWriter.cs

```diff
@@ -0,0 +1,217 @@
+            var eventQueue = writer._eventQueue;
+            var completionSource = writer._flushTaskCompletionSource;
+            var flushDelayEvent = writer._flushDelayEvent;
+            var ciTestCycleBuffer = writer._ciTestCycleBuffer;
+
+            Log.Debug("CIAgentlessWriter:: InternalFlushEventsAsync/ Starting FlushEventsAsync loop");
+
+            while (!eventQueue.IsCompleted)
+            {
+                TaskCompletionSource<bool> watermarkCompletion = null;
+
+                try
+                {
+                    // Retrieve events from the queue and add them to the respective buffer.
+                    while (eventQueue.TryTake(out var item))
+                    {
+                        if (item is WatermarkEvent watermarkEvent)
+                        {
+                            // Flush operation.
+                            // We get the completion source and exit this loop
+                            // to flush buffers (in case there's any event)
+                            watermarkCompletion = watermarkEvent.CompletionSource;
+                            break;
+                        }
+                        else if (ciTestCycleBuffer.CanProcessEvent(item))
+                        {
+                            // The CITestCycle endpoint can process this event, we try to add it to the buffer.
+                            if (!ciTestCycleBuffer.TryProcessEvent(item) && ciTestCycleBuffer.HasEvents)
+                            {
+                                // If the item cannot be added to the buffer but the buffer has events
+                                // we assume that is full and needs to be flushed.
+                                await writer.SendPayloadAsync(ciTestCycleBuffer).ConfigureAwait(false);
+                                ciTestCycleBuffer.Clear();
+                            }
+                        }
+                    }
+
+                    // After removing all items from the queue, we check if buffers needs to flushed.
+                    if (ciTestCycleBuffer.HasEvents)
+                    {
+                        // flush is required
+                        await writer.SendPayloadAsync(ciTestCycleBuffer).ConfigureAwait(false);
+                        ciTestCycleBuffer.Clear();
+                    }
+
+                    // If there's a flush watermark we marked as resolved.
+                    watermarkCompletion?.TrySetResult(true);
+                }
+                catch (ThreadAbortException ex)
+                {
+                    completionSource?.TrySetException(ex);
+                    watermarkCompletion?.TrySetException(ex);
+                    throw;
+                }
+                catch (Exception ex)
+                {
+                    watermarkCompletion?.TrySetException(ex);
+                }
+                finally
+                {
+                    if (watermarkCompletion is null)
```

Comment:
I do not know the whole mechanism, sorry in advance:
What if we flushed events but there was no watermark and no body calls the `FlushTracesAsync()` method.
The `eventQueue` will grow indefinitely not ?

Reply:
not exactly, we are flushing every second. and also we are using a bounded blocking collection. In a CI scenario we rather prefer to stop the world and wait for flush than drop events.

Reply:
Thanks for the explanation